### PR TITLE
Move `DifficultyControlPoint`s to be specified at a per-`HitObject` level

### PR DESCRIPTION
**Pull request by peppy**
SImilar change to the recent `SampleControlPoint` refactor. I'm still testing this for any signs for regressions, but is in a reviewable state.

### CODE REVIEW

File: osu.Game/Beatmaps/BeatmapDifficulty.cs
Comment (bdach) [inline osu.Game/Beatmaps/BeatmapDifficulty.cs:29]: Stable internally seems to use `DifficultySliderMultiplier`, incidentally, which I like better than this. But that's for later.

File: osu.Game/Rulesets/Edit/HitObjectComposer.cs
Comment (bdach) [inline osu.Game/Rulesets/Edit/HitObjectComposer.cs:393]: I have a slight feeling that doing this may break scrolling rulesets as those seem to be using the effect points' `ApproachRate` thing. Haven't tested, though.

File: osu.Game/Rulesets/Edit/IPositionSnapProvider.cs
Comment (bdach) [inline osu.Game/Rulesets/Edit/IPositionSnapProvider.cs:31]: xmldocs not filled out

File: osu.Game/Screens/Edit/Timing/DifficultySection.cs
Comment (bdach) [inline osu.Game/Screens/Edit/Timing/DifficultySection.cs:22]: Kind of a drive-by place to put it but effect points probably need UI for specifying approach rate, too, at some point. Although I'm not sure how to best explain what the difference is (I'm not entirely sure I have a good grasp on it myself yet).

**Review by bdach — DISMISSED** [review]

File: osu.Game.Rulesets.Mania.Tests/TestSceneHoldNoteInput.cs
Comment (peppy) [reply osu.Game.Rulesets.Mania.Tests/TestSceneHoldNoteInput.cs:391]: Yep, for sure. Some of these tests are unfortunately visual only, and as per my own admission I haven't gone through and tested all scenarios yet.

File: osu.Game/Beatmaps/BeatmapDifficulty.cs
Comment (peppy) [reply osu.Game/Beatmaps/BeatmapDifficulty.cs:29]: The hope would be to remove this completely, as it should only really be required as an editor "map default" if anything, once we're done.

**Comment by peppy** [comment]
Hold off on further review on this one asI have a few discussion points to resolve first.

In summary:

- should AR be global adjust (ie osu mode) rather than per object?
- should effect point creation be moved to converter (rationale: osu doesn’t want)
- shall we remove difficultycontrolpoint and replace with just SV?

**Comment by peppy** [comment]
I've made initial changes in line with discussion points above. For now I've decided to not remove/refactor `DifficultyControlPoint` itself as there's still quite a few non-osu! ruleset accesses to the value. Probably best left to be reconsidered in a future PR.

File: osu.Game/Beatmaps/BeatmapDifficulty.cs
Comment (peppy) [reply osu.Game/Beatmaps/BeatmapDifficulty.cs:29]: You mean just the prefix? I think it's about the same since this whole class is difficulty-grouped attributes of the beatmap.

Comment (bdach) [reply osu.Game/Beatmaps/BeatmapDifficulty.cs:29]: Yeah now that the other instances of "slider multiplier" have been culled it may be fine to leave this be as no longer ambiguous.

File: osu.Game.Rulesets.Mania.Tests/TestSceneHoldNoteInput.cs
Comment (peppy) [reply osu.Game.Rulesets.Mania.Tests/TestSceneHoldNoteInput.cs:391]: I'm doing a second pass on this be actually throwing when a control point is attempted to be added to a `ControlPointInfo` that doesn't handle its type. Should hopefully catch all these cases (and has initially surfaced quite a few new cases to check on).

File: osu.Game/Rulesets/Timing/MultiplierControlPoint.cs
Comment (peppy) [reply osu.Game/Rulesets/Timing/MultiplierControlPoint.cs:41]: The general gist is that:

- `ApproachRate` (now renamed to `ScrollSpeed`) is what handles global scroll speed adjustments
- `SliderVelocity` is what handles per `HitObject` slider velocity (and is only read after being attached to hitobjects with this change)

It's mostly an organisational change, where control points are only intended to be used going forward for applications which can't be applied directly to hitobjects. In an attempt to make things more understandable and better allow the underlying data structure to match the expected UX of the editor.

File: osu.Game/Screens/Edit/Timing/DifficultySection.cs
Comment (peppy) [reply osu.Game/Screens/Edit/Timing/DifficultySection.cs:22]: Has been added.

File: osu.Game.Rulesets.Mania.Tests/TestSceneHoldNoteInput.cs
Comment (peppy) [reply osu.Game.Rulesets.Mania.Tests/TestSceneHoldNoteInput.cs]: @bdach can you explain the `DistanceObjectPatternGenerator` one? Is there test coverage of this somewhere i'm missing, or do you mean the usage in the class itself? If the second, I *think* that's actually required to get some weird precision from the special legacy point type.

Comment (bdach) [reply osu.Game.Rulesets.Mania.Tests/TestSceneHoldNoteInput.cs]: Well I mean it seems a bit weird that almost *all* of mania changes in this diff use `ScrollSpeed` except for this one case that uses `SliderVelocity`. Especially so that `ManiaBeatmapConverter` converts slider velocity adjustments and difficulty control points to scroll speed adjustments and effect control points in the ctor, and that happens before the pattern generators are invoked I'm pretty sure...

`ManiaBeatmapConverterTest` exists but from a quick look I don't think it tests nonstandard scroll speeds in the context of the stair pattern generation.

The precision stuff is there but it may be required to make the `LegacyDifficultyControlPoint` a `LegacyEffectControlPoint` now?

File: osu.Game/Rulesets/Edit/HitObjectComposer.cs
Comment (peppy) [reply osu.Game/Rulesets/Edit/HitObjectComposer.cs:393]: This is actually a member of `IPositionSnapProvider`, which is only used by osu! for positional / distance snap. So while it looks a bit weird being in such a central class as `HitObjectComposer`, it's correct implementation.

File: osu.Game.Rulesets.Mania.Tests/TestSceneHoldNoteInput.cs
Comment (peppy) [reply osu.Game.Rulesets.Mania.Tests/TestSceneHoldNoteInput.cs]: For sure. Going to go over this with a fine comb. I was trying to avoid changing/dealing with this because I still don't even fully understand it, but let's see how things go..

Comment (peppy) [reply osu.Game.Rulesets.Mania.Tests/TestSceneHoldNoteInput.cs]: So following the flow, this generation pass is done *after* the conversion of `DifficultyControlPoint`s to `EffectControlPoint`s... but the conversion is only done for osu!mania specific beatmaps, which would mean there is no object conversion.

So in a round-about convoluted way, this is actually correct as it stands. Not sure how I feel about that in terms of being able to understand everything going on. @smoogipoo can you chime in on this one when you get to reviewing this PR?

Comment (bdach) [reply osu.Game.Rulesets.Mania.Tests/TestSceneHoldNoteInput.cs]: Oh, right, that explains the difference in handling. I'd personally be ok with an inline comment in that case.

**Comment by bdach** [comment]
Have re-read this diff top to bottom today and it seems clearer than the previous time around. That might be just the familiarity, though. Either way seems good to me. :+1: 

Haven't tested this in any way, shape or form, mind. I'm mostly hoping test coverage is sufficient.

**Comment by smoogipoo** [comment]
!pp check 0 1 2 3

**Comment by smoogipoo** [comment]
Diffcalc run will be at https://github.com/smoogipoo/osu/actions/runs/1216065798

File: osu.Game/Beatmaps/ControlPoints/DifficultyControlPoint.cs
Comment (smoogipoo) [inline osu.Game/Beatmaps/ControlPoints/DifficultyControlPoint.cs:23]: It seems weird to call this "SliderVelocity" when it's something fundamental to hitobjects and should probably be named in a general way as such.

What's the path forward here? Is it to eventually remove `DifficultyControlPoint` and give `Slider` a "Velocity" property, or is it to continue using `DifficultyControlPoint` for hitobjects? If the latter, then I propose this to be renamed to just "Velocity".

**Review by smoogipoo — CHANGES_REQUESTED** [review]
The rest looks fine on a code pass.

File: osu.Game/Beatmaps/ControlPoints/DifficultyControlPoint.cs
Comment (peppy) [reply osu.Game/Beatmaps/ControlPoints/DifficultyControlPoint.cs:23]: yeah, next step is to nuke this completely. can rename i guess, though much of a muchness. i feel SliderVelocity gave a better idea of direction?

File: osu.Game.Rulesets.Mania/Beatmaps/ManiaBeatmapConverter.cs
Comment (peppy) [reply osu.Game.Rulesets.Mania/Beatmaps/ManiaBeatmapConverter.cs:70]: I guess we also need test coverage against this kind of modification?

Comment (peppy) [reply osu.Game.Rulesets.Mania/Beatmaps/ManiaBeatmapConverter.cs]: I've added test coverage for this kind of failure.

Hopefully you're okay with the extra line added to `convert-samples` to get a fail (there was no existing mania-specific beatmap with inherited timing data... can potentially add a dedicated one if that's preferred)

File: osu.Game/Beatmaps/ControlPoints/DifficultyControlPoint.cs
Comment (smoogipoo) [reply osu.Game/Beatmaps/ControlPoints/DifficultyControlPoint.cs:23]: I think it's fine for now in that case.

**Comment by smoogipoo** [comment]
No diffcalc differences as of the current state of this PR.

**Comment by peppy** [comment]
Tentatively pushed changes here. Needs some careful review consideration because I'm still not 100% sure. My goal is to avoid breaking anything while clearing a path forward for removing the `ControlPoint`s from hit objects completely.

General:
- Added support for nested hitobjects. Until now these were being ignored (which probably says we need more compehensive tests -- I believe cases mentioned at https://github.com/ppy/osu/pull/3815 would be broken, ie. where there are `SampleControlPoints` only affecting certain ticks).

For editor flow:
- Cloned out the control points when storing to hitobjects. This is required to have the editor and re-encoding work correctly. Keep in mind this logic should be removed after we remove the control points. Hopefully it's not too bad for the time being.
- Stored the time data expected to be used for lookups in the cloned control points. Without this, multiple control points may have had a reference to the same legacy point. If an earlier one was modified in the editor (even via replacing with a new instance), later objects would have control points with an earlier time, causing them to overwrite the modified control points.


With these changes it is feasible to use the editor and load / save SV changes applied directly to hitobjects. You can test the editor portion on https://github.com/ppy/osu/compare/master...peppy:slider-timeline-velcotiy-adjust-v2?expand=1.

While the "editor flow" points may seem like they could be moved into `EditorBeatmap` (along with the legacy control point removal in the linked branch). Unfortunately this isn't enough as the editor uses legacy encode/decode for diff and change handling.

Change SV and save / load:

https://user-images.githubusercontent.com/191335/132827759-096799ac-090d-48e4-8e2d-0663533a3e68.mp4

Place new slider at "current" SV (note different speeds):

https://user-images.githubusercontent.com/191335/132827876-df7a923a-a362-415b-b1c4-8944605b940a.mp4

Load into stable:

20210910 175336 (Parallels Desktop app)

Note that playback in stable is currently broken due to floating point precision. Still investigating this, but should be considered out of scope of this PR.

File: osu.Game/Rulesets/Objects/HitObject.cs
Comment (bdach) [inline osu.Game/Rulesets/Objects/HitObject.cs:102]: As a side note we may eventually have to have `EndTimeBindable` to fix issues like https://github.com/ppy/osu/issues/13473. Out of scope here, just mentioning in passing.

Also, on comparing with `ApplyDefaults()` lower down, is there a reason why `control_point_leniency` is dropped here? I guess the assumption is that start time should be snapped to full milliseconds or something?

File: osu.Game/Beatmaps/Formats/LegacyBeatmapEncoder.cs
Comment (bdach) [inline osu.Game/Beatmaps/Formats/LegacyBeatmapEncoder.cs:188]: Maybe not super relevant, but I'll mention it: the way these are implemented sort of breaks down with concurrent objects. Consider something like the following case:

```
-------------------------------------> time

 [x]-----[o]-----[o]-----[o]-----[o]
     [x]
             [x]
                     [x]
```

where `[x]` signifies a hitobject's start time, and `[o]` signified a nested's start time. This procedure as written is silently relying on the fact that there are no other simultaneous hitobjects while an object with nested is being processed.

For full correctness you'd probably have to collect all hitobjects to a flat list, sort by start time and only then process. (This still doesn't fully solve a potential issue with hitobjects with duration, wherein you may have to "restore" the previous control point after processing another concurrent one, but that may be out of scope for now.)

This will only be relevant when encoding a beatmap with per-object points back into the legacy format. A new format that doesn't attempt to serialise back to the old format will not have such issues.

File: osu.Game/Rulesets/Objects/HitObject.cs
Comment (peppy) [reply osu.Game/Rulesets/Objects/HitObject.cs:102]: i'm not sure which is correct at this point. hard to say without some intensive testing, which might be out of scope for now..

**Comment by smoogipoo** [comment]
Retested again, still no diffcalc differences.

File: osu.Game/Beatmaps/ControlPoints/ControlPoint.cs
Comment (peppy) [inline osu.Game/Beatmaps/ControlPoints/ControlPoint.cs:18]: @smoogipoo during your re-review, this may be a line of note.

rationale behind this being okay is that we never update group's `Time` (and `ControlPoints` are always `AttachGroup`ed), so they can't get out of sync.

File: osu.Game/Beatmaps/Formats/LegacyBeatmapEncoder.cs
Comment (peppy) [inline osu.Game/Beatmaps/Formats/LegacyBeatmapEncoder.cs:261]: not sure this comment is clear enough. basically the sample points need to be processed with nested first to avoid nested sample data being lost due to the presence of the hitobject itself. reversing the order makes tests fail which is how i got here.

File: osu.Game/Rulesets/Objects/HitObject.cs
Comment (peppy) [reply osu.Game/Rulesets/Objects/HitObject.cs:102]: This may be the reason for exports to stable not getting correct velocities applied, and needs further investigation before merge.

**Comment by peppy** [comment]
Still welcoming review here, but going to add blocked label until I resolve the stable import issue mentioned in https://github.com/ppy/osu/pull/14619#issuecomment-916748972.

File: osu.Game/Rulesets/Objects/HitObject.cs
Comment (smoogipoo) [reply osu.Game/Rulesets/Objects/HitObject.cs:102]: As I mentioned before, end time is likely more correct here as hit samples are played when a hitobject ends and not when it starts. That being said I forgot how slider slide/whistle/spinner samples work.

Is there a reason for changing this? Isn't it easy to `GetEndTime() + control_point_leniency` as before?

I would rather not change things because "i'm not sure which is correct at this point" when we've resolved many issues over the years to get hitsounds on-par with stable.

Comment (peppy) [reply osu.Game/Rulesets/Objects/HitObject.cs:102]: Yep, I'll definitely try changing it back. The issue is that the only part which this affects is playback on stable, mostly in cases which we haven't encountered and don't have tests for yet.

Picture the scenario where in lazer, a mapper assigns a sample to a tick, which is encoded out to be `tick.EndTime+allowance` as a global `SampleControlPoint`. It's feasible that this could be read incorrectly by stable, potentially applying to the hitobject itself (if it occurs towards to end time of the parent object) I think? May also be why stable uses different values for the lenience for different lookups.

Either way, this is part of the remaining tests that are required to get this PR merged. A similar issue is confirmed to already be occurring for SV changes, as I've mentioned below.

**Comment by peppy** [comment]
> Still welcoming review here, but going to add blocked label until I resolve the stable import issue mentioned in [#14619 (comment)](https://github.com/ppy/osu/pull/14619#issuecomment-916748972).

This is happening due to osu-stable actually parsing control point times as `double`s. After further testing and discussion we've agreed to change stable to cast to `int`, to keep things simple. The editor doesn't allow any kind of floating point precision to be specified, so the only cases where this could possible have an effect is if a user was to have altered the `.osu` file directly. If we ever come across a case like this we can fix that manually (one would hope there are none that are ranked).

**Comment by peppy** [comment]
One thing to note (which I did touch on) is that slider ticks will actually potentially not work as expected with the current logic. This is because the leniency specification for stable lookups is different to that of the main hitobject:

20210914 170105 (Parallels Desktop)

Not sure if this is out of scope for this PR, but we will probably want to handle this properly at some point. Once we allow a user to assign samples directly to ticks / nested objects, the `5ms` leniency we are currently adding will cause them to be one tick behind where they should be.

**Comment by bdach** [comment]
I'm not sure where this PR is at exactly, but as a way of testing this change, I've jerry-rigged a trimmed version of one of the tests added here (`LegacyBeatmapEncoderTest.TestEncodeDecodeStabilityWithNonLegacyControlPoints`) to run against the latest dump of beatmaps from September from https://data.ppy.sh to see if that test would fail on any of those maps.

At the very start, I had that test fail on 12051 maps; I have thus far managed to reduce that set down to 8188 maps, and have another change that will cut that number down further, but progress on that is slow because it's quite an arduous and semi-manual process as I have it. I may need a few days more to get that number down to zero.

File: osu.Game/Beatmaps/Formats/LegacyBeatmapDecoder.cs
Comment (bdach) [inline osu.Game/Beatmaps/Formats/LegacyBeatmapDecoder.cs:383]: One thing that came out of the tests above is - for scrolling rulesets, should `speedMultiplier` not be stored to the `EffectControlPoint` instead to have the equivalent effect? An example of a map that this caused a failure on is https://osu.ppy.sh/beatmapsets/460422#fruits/1000293, for instance.

(Incidentally that also raises the issue of the precision of that number, as the effect point's `ScrollSpeedBindable`'s precision is 0.01, rather than the epsilon that the legacy point used to use, but that might not be super relevant.)

**Comment by peppy** [comment]
> I'm not sure where this PR is at exactly

It's intended to be in a mergeable state.

> so I may need a few days more to get that number down to zero.

I'll block for now. Sounds like quite an invaluable test to have available to us going forward (marked as `[Ignore]` or something so we can just activate as a manual run).

**Comment by bdach** [comment]
> Sounds like quite an invaluable test to have available to us going forward (marked as `[Ignore]` or something so we can just activate as a manual run).

To be honest I wasn't really planning on adding that test, as it's really rough around the edges, due to being hampered by some limitations (as it turns out rider/nunit can't really deal with over 100k test cases for a test, so it runs as one big blob now) and also using real OS paths for reads and writes (I am writing out the full serialised beatmaps to a directory to compare later). But I'll see if I can clean it up at the end of the process.

**Comment by peppy** [comment]
I’d see it being a single test which iterates over files in a directory, or something like that.

**Comment by bdach** [comment]
As an update, I think I identified all failures I was seeing and I have what I think are explanations (and fixes, where relevant) for them. I want to run all of my changes once again over the entire set overnight to make sure I didn't break any cases that were working before, but the preliminary results are [here](https://gist.github.com/bdach/ce54974790f145aba68cf5fd42e954a2).

**Comment by bdach** [comment]
After re-running on the full set, I did not identify any more failing maps than the 1110 that I already know to be failing for reasons unrelated to this pull. So aside from the two changes proposed in the gist above I can't really think of anything else that I can do to verify this further.

As for including the test that I was doing this on, I think I'd be looking to PR it separately to not inflate this diff more than it already is.

**Comment by peppy** [comment]
Feel free to commit the proposed changes to this branch. Not too sure about the commented out precision change (did you explain this anywhere or was it just a "works better without" change?) but as long as it passes difficulty calculation re-testing then it shouldn't be an issue. I believe that may be the main reason for it having existed, so it'll be interesting to see.

**Comment by bdach** [comment]
The precision change was just to temporarily silence changes like scroll speed going from 1.499999999 to 1.5 (due to differences in precision between legacy difficulty points and effect points) showing up on diffs. It's not something I'd ever want to be caught seriously committing.

**Comment by peppy** [comment]
Right, that makes more sense.

**Comment by bdach** [comment]
I've tentatively pushed my changes. Particularly of note is 8026968, which is a different way of fixing one of the issues than I applied in my report above - it's pretty much a revert of 3c7a34b, because the difficulty -> effect point conversion wrt scroll speed is not specific to mania, but rather to all scrolling rulesets.

Do note that this means that both the decoder and encoder now use `isOsuRuleset` as essentially a "is this a scrolling ruleset" flag. It may be fine to do this for now, or not. I'm in two minds, but I'm going with what the encoder is doing for now. Can change back to doing the switch in converters instead.

Not unblocking yet because I want to re-run the full set of beatmaps again overnight, but I've tested the three concrete examples that I debugged and they pass, so I'm relatively certain that even these new changes should fix the issues I found.

**Comment by bdach** [comment]
I broke diffcalc with my changes because catch (`JuiceStream` in particular) is still reading slider velocity from difficulty points, while I thought all scrolling rulesets should now be using effect points (because of the encoding logic, and because effect points is what `MultiplierControlPoint` reads, which in turn controls scrolling containers). No longer sure that should be the case?

**Comment by bdach** [comment]
Have re-tested overnight and still at the same number of beatmaps broken (1110, all irrelevant to these changes), so going to remove the blocked tag.

**Comment by peppy** [comment]
You bring up a good point regarding `JuiceStream`. It actually looks wrong as it stands. The way it's currently written sees the `Velocity` and `TickDistance` being decided by `SliderVelocity` with the intention of matching the scroll rate, but the scroll rate is coming from `EffectControlPoint.ScrollSpeed`.

Is it feasible that we can change the logic used there and reapply your reverted change?

Something like:

```diff
diff --git a/osu.Game.Rulesets.Catch/Objects/JuiceStream.cs b/osu.Game.Rulesets.Catch/Objects/JuiceStream.cs
index cb8056231c..e93da7d8a1 100644
--- a/osu.Game.Rulesets.Catch/Objects/JuiceStream.cs
+++ b/osu.Game.Rulesets.Catch/Objects/JuiceStream.cs
@@ -43,7 +43,7 @@ protected override void ApplyDefaultsToSelf(ControlPointInfo controlPointInfo, B
 
             TimingControlPoint timingPoint = controlPointInfo.TimingPointAt(StartTime);
 
-            double scoringDistance = base_scoring_distance * difficulty.SliderMultiplier * DifficultyControlPoint.SliderVelocity;
+            double scoringDistance = base_scoring_distance * difficulty.SliderMultiplier * controlPointInfo.EffectPointAt(StartTime).ScrollSpeed;
 
             Velocity = scoringDistance / timingPoint.BeatLength;
             TickDistance = scoringDistance / difficulty.SliderTickRate;
```

**Comment by bdach** [comment]
I believe that would indeed be feasible, but I think the taiko usages would need to receive the same treatment as well (although the diffcalc test for taiko pass either way, so they're likely just not covering scroll speed).

**Comment by peppy** [comment]
!pp check osu taiko catch mania

**Comment by peppy** [comment]
!pp check osu taiko catch mania

**Comment by bdach** [comment]
The diffcalc run [failed](https://github.com/ppy/osu/actions/runs/1298136563) due to exceeding the default runtime limit of 6 hours. Self-hosted runners can apparently go up to [72 hours](https://docs.github.com/en/actions/hosting-your-own-runners/about-self-hosted-runners#usage-limits), although given that the mania run alone seems to have processed just above 10% of everything that is likely to not be enough either...

**Comment by peppy** [comment]
> The diffcalc run [failed](https://github.com/ppy/osu/actions/runs/1298136563) due to exceeding the default runtime limit of 6 hours. Self-hosted runners can apparently go up to [72 hours](https://docs.github.com/en/actions/hosting-your-own-runners/about-self-hosted-runners#usage-limits), although given that the mania run alone seems to have processed just above 10% of everything that is likely to not be enough either...

It got through 10% of the second run, so that should actually buy us enough leeway for now. I'll give it a try, didn't realise it could go higher for self runners.

We also discussed not necessarily requiring to run with converts, but I'm not sure a decision has been made regarding that yet.

**Comment by peppy** [comment]
!pp check osu taiko catch mania

**Comment by smoogipoo** [comment]
A few test failures. Maybe caused by my merge but unsure.

**Comment by bdach** [comment]
counterintuitively the CI failure is caused by a test added in *this* PR. taiko's slider velocity stuff is modifying the beatmap being converted. so in fact it is `master` that is wrong/buggy here.

i'll look into fixing that separately from this PR.

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**
